kernel/work.c

```diff
@@ -0,0 +1,1030 @@
+ * the queue it was submitted to.  That may or may not be the queue provided
+ * on input.
+ *
+ * @retval 0 if already queued
+ * @retval 1 if unqueued and not running
+ * @retval 2 if unqueued and running
+ * @retval -EBUSY if canceling or submission was rejected by queue
+ * @retval -EINVAL if no queue is provided
+ * @retval -ENODEV if the queue is not started
+ */
+static int submit_to_queue_locked(struct k_work *work,
+				  struct k_work_q **queuep)
+{
+	int ret = 0;
+
+	if (flag_test(&work->flags, K_WORK_CANCELING_BIT)) {
+		/* Disallowed */
+		ret = -EBUSY;
+	} else if (!flag_test(&work->flags, K_WORK_QUEUED_BIT)) {
+		/* Not currently queued */
+		ret = 1;
+
+		/* If no queue specified resubmit to last queue.
+		 */
+		if (*queuep == NULL) {
+			*queuep = work->queue;
+		}
+
+		/* If the work is currently running we have to use the
+		 * queue it's running on to prevent handler
+		 * re-entrancy.
+		 */
+		if (flag_test(&work->flags, K_WORK_RUNNING_BIT)) {
+			__ASSERT_NO_MSG(work->queue != NULL);
+			*queuep = work->queue;
+			ret = 2;
+		}
+
+		int rc = queue_submit_locked(*queuep, work);
+
+		if (rc < 0) {
+			ret = rc;
+		} else {
+			flag_set(&work->flags, K_WORK_QUEUED_BIT);
+			work->queue = *queuep;
+		}
+	} else {
+		/* Already queued, do nothing. */
+	}
+
+	if (ret <= 0) {
+		*queuep = NULL;
+	}
+
+	return ret;
+}
+
+int k_work_submit_to_queue(struct k_work_q *queue,
+			    struct k_work *work)
+{
+	__ASSERT_NO_MSG(work != NULL);
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	int ret = submit_to_queue_locked(work, &queue);
+
+	k_spin_unlock(&lock, key);
+
+	/* If we changed the queue contents (as indicated by a positive ret)
+	 * the queue thread may now be ready, but we missed the reschedule
+	 * point because the lock was held.  If this is being invoked by a
+	 * preemptible thread then yield.
+	 */
+	if ((ret > 0) && (k_is_preempt_thread() != 0)) {
+		k_yield();
+	}
+
+
+	return ret;
+}
+
+/* Flush the work item if necessary.
+ *
+ * Flushing is necessary only if the work is either queued or running.
+ *
+ * Invoked with work lock held by key.
+ * Sleeps.
+ *
+ * @param work the work item that is to be flushed
+ * @param flusher state used to synchronize the flush
+ * @param key the key under which the work lock is held.
+ *
+ * @retval true if work is queued or running.  If this happens the
+ * caller must take the flusher semaphore after releasing the lock.
+ *
+ * @retval false otherwise.  No wait required.
+ */
+static bool work_flush_locked(struct k_work *work,
+			      struct z_work_flusher *flusher)
+{
+	struct k_work_q *queue = NULL;
+	bool need_flush = (flags_get(&work->flags)
+			   & (K_WORK_QUEUED | K_WORK_RUNNING)) != 0;
+
+	if (need_flush) {
+		queue = work->queue;
+		__ASSERT_NO_MSG(queue != NULL);
+
+		queue_flusher_locked(queue, work, flusher);
+		notify_queue_locked(queue);
+	}
+
+	return need_flush;
+}
+
+bool k_work_flush(struct k_work *work,
```

Comment:
That concern also applies to `k_work_{,delayable_}cancel_sync`.

The documented behavior is that the call doesn't return until the work has completed (cancellation).  If the thread that's invoking the operation is the thread that will signal completion of the operation, yes, it'll deadlock.  Just as could happen with any semaphore both given and taken by the same thread.

We could change the API so that it can also return an error code saying that the cancel/flush operation was rejected, if this is a serious concern.

Otherwise I think the answer is "Don't do that".  I've added another note to the documentation.